File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.util.*;

public class Main
{
    public static int N;
    public static int S;
    public static int[] input;
    public static int answer = 0;
	public static void main(String[] args) {
	    Scanner scanner = new Scanner(System.in);
	    N = scanner.nextInt();
	    S = scanner.nextInt();
	    
	    input = new int[N];
	    for(int i = 0; i < N; i++) {
	        input[i] = scanner.nextInt();
	    }
	    
	    DFS(0, 0);
	    
	    if(S == 0) {
	        answer--;
	    }
	    
	    System.out.println(answer);
	}
	
	public static void DFS(int depth, int sum) {
	    if(depth == N) {
	        if(sum == S) {
	            answer++;
	            return;
	        }
	        return;
	    }
	    
	    DFS(depth + 1, sum);
	    DFS(depth + 1, sum + input[depth]);
	}
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.py
import itertools


def is_target_number(numbers, target_number):
    total_count = 0

    # 부분 수열의 길이
    for i in range(1, len(numbers) + 1):

        # i길이 만큼의 부분수열 생성
        for sub_numbers in itertools.combinations(numbers, i):
            sum_sub_numbers = sum(sub_numbers)

            if sum_sub_numbers == target_number:
                total_count = total_count + 1

    return total_count


N, S = map(int, input().split())

numbers = list(map(int, input().split()))

answer = is_target_number(numbers, S)

print(answer)